src/controllers/authenticationController.ts

```diff
@@ -21,74 +21,138 @@ export default class AuthenticationController {
    */
   // tslint:disable-next-line:no-any
-  public acs(userPayload: any, res: express.Response): void {
-    winston.log(
-      "info",
-      "Called assertion consumer service with data:",
-      userPayload
-    );
-
+  public async acs(userPayload: any): Promise<Either<Error, IResponse>> {
     const errorOrUser = validateSpidUser(userPayload);
 
-    errorOrUser.fold(
-      (error: Error) => this.return500Error(error, res),
-      (spidUser: SpidUser) => {
-        const user = toAppUser(spidUser, this.tokenService.getNewToken());
+    if (isLeft(errorOrUser)) {
+      const error = errorOrUser.value;
+      return left(error);
+    }
 
-        this.sessionStorage.set(user.token, user);
+    const spidUser = errorOrUser.value;
+    const user = toAppUser(spidUser, this.tokenService.getNewToken());
 
-        const url =
-          process.env.CLIENT_REDIRECTION_URL || "/profile.html?token={token}";
-        const urlWithToken = url.replace("{token}", user.token);
+    const timestamp = Date.now();
+    const errorOrBool = await this.sessionStorage.set(
+      user.token,
+      user,
+      timestamp
+    );
 
-        res.redirect(urlWithToken);
-      }
+    if (isLeft(errorOrBool)) {
+      const error = errorOrBool.value;
+      return left(error);
+    }
+    const response = errorOrBool.value;
+
+    if (!response) {
+      return left(new Error("Error creating the user session"));
+    }
+    const urlWithToken = this.clientProfileRedirectionUrl.replace(
+      "{token}",
+      user.token
     );
+
+    return right({
+      body: urlWithToken,
+      status: 301
+    });
   }
 
   /**
    * Retrieves the logout url from the IDP.
    */
-  public logout(req: express.Request, res: express.Response): void {
+  public async logout(req: express.Request): Promise<Either<Error, IResponse>> {
     const errorOrUser = extractUserFromRequest(req);
 
-    errorOrUser.fold(
-      (error: Error) => this.return500Error(error, res),
-      (user: User) => {
-        // Delete the Redis token.
-        this.sessionStorage.del(user.token);
-
-        // Logout from SPID.
-        req.query = {};
-        req.query.entityID = user.spid_idp;
-
-        this.spidStrategy.logout(req, (err, request: express.Request) => {
-          if (!err) {
-            res.status(200).json({
-              logoutUrl: request
-            });
-          } else {
-            res.status(500).json({
-              message: err.toString()
-            });
-          }
-        });
+    if (isLeft(errorOrUser)) {
+      const error = errorOrUser.value;
+      return left(error);
+    }
+
+    const user = errorOrUser.value;
+
+    const errorOrBool = await this.sessionStorage.del(user.token);
+
+    if (isLeft(errorOrBool)) {
+      const error = errorOrBool.value;
+      return left(error);
+    }
+
+    const response = errorOrBool.value;
+
+    if (!response) {
+      return left(new Error("Error creating the user session"));
+    }
+
+    // Logout from SPID.
```

Comment:
this is a bit obscure, could you explain what it's happening here in a short comment, perhaps linking the relevant docs?